README
SDM4 importer til sks

Changelog.md
## sksimporter 4.3
*  Opgrading til sdm4-core 4.3, der løser
   NSPSUPPORT-126: ParserExecutor logger filers absolutte stier og md5-summer inden parser behandler dem
*  Opgraderet til sdm-core 4.9 for at kunne resette transaction time
*  SDM-23: Reset transaction time før hver import

## sksimporter 4.4
*  Opdateret SDM4 depencencies
*  SDM-5 SLA-log fra SDM4-importere følger ikke standarden
*  Tilføjet kopi register view, så kopi register service maps nu bliver oprettet automatisk

## sksimporter 4.5
*  Rettet: SDMSKSIMP-1 SKS importeren misfortolker slut dato


src/test/java/dk/nsi/sdm4/sks/SksParserTest.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dk.nsi.sdm4.sks;

import org.apache.commons.io.FileUtils;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import java.io.File;
import java.io.IOException;

import static org.apache.commons.io.FileUtils.toFile;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class SksParserTest
{
    private static String completeTxt = "data/sks/SHAKCOMPLETE.TXT";
    private static String completeXml = "data/sks/SHAKCOMPLETE.XML";
    private static String delta = "data/sks/SHAKDELTA.TXT";

	private SKSParser importer;

	@Rule
	public TemporaryFolder tmpDir = new TemporaryFolder();

	@Before
	public void setup() {
		importer = new SKSParser();
	}

	@Test(expected = IllegalArgumentException.class)
    public void shouldNotAcceptEmptyInputFileSet() throws IOException {
        importer.validateInputStructure(tmpDir.newFolder());
    }
    
    @Test
    public void shouldAcceptSKSCompleteTxtFile() throws IOException {
        assertTrue(importer.validateInputStructure(datasetDirWith(completeTxt)));
    }

	@Test
    public void shouldNotAcceptSKSCompleteXmlFile() throws IOException {
        assertFalse(importer.validateInputStructure(datasetDirWith(completeXml)));
    }
    
    @Test
    public void shouldAcceptSKSDeltaFile() throws IOException {
        assertTrue(importer.validateInputStructure(datasetDirWith(delta)));
    }

	private File datasetDirWith(String filename) throws IOException {
		File datasetDir = tmpDir.newFolder();
		FileUtils.copyFileToDirectory(getFile(filename), datasetDir);
		return datasetDir;
	}

	private File getFile(String filename) {
		return toFile(getClass().getClassLoader().getResource(filename));
	}
}


src/test/java/dk/nsi/sdm4/sks/SksIntegrationTest.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package dk.nsi.sdm4.sks;

import dk.nsi.sdm4.sks.config.SksimporterApplicationConfig;
import dk.nsi.sdm4.testutils.TestDbConfiguration;
import org.apache.commons.io.FileUtils;
import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormatter;
import org.joda.time.format.ISODateTimeFormat;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
import org.springframework.transaction.annotation.Transactional;

import java.io.File;
import java.io.IOException;
import java.sql.Timestamp;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

import static org.apache.commons.io.FileUtils.toFile;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;


@RunWith(SpringJUnit4ClassRunner.class)
@Transactional
@ContextConfiguration(classes = {SksimporterApplicationConfig.class, TestDbConfiguration.class})
public class SksIntegrationTest
{
	@Rule
	public TemporaryFolder tmpDir = new TemporaryFolder();

	@Autowired
	JdbcTemplate jdbcTemplate;

	@Autowired
	SKSParser importer;

	@Test
	public void canImportTheCorrectNumberOfRecords() throws Throwable {
		importer.process(datasetDirWith("data/sks/SHAKCOMPLETE.TXT"), "");
		
		// FIXME: These record counts are only correct iff if duplicate keys are disregarted.
		// This is unfortunate. Keys are currently only considered based their SKSKode.
		// They should be a combination of type + kode + startdato based on the register doc.
		assertEquals(622, jdbcTemplate.queryForInt("SELECT COUNT(*) FROM Organisation WHERE Organisationstype = 'Sygehus'"));
		assertEquals(8451, jdbcTemplate.queryForInt("SELECT COUNT(*) FROM Organisation WHERE Organisationstype = 'Afdeling'"));
	}

    @Test
    public void validToAndFromInclusive() throws IOException, ParseException {
        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        // Rigshospitalet have the following date specified in input:
        // Valid from inclusive = 19760401
        // Valid to inclusive = 25000101

        importer.process(datasetDirWith("data/sks/SHAKCOMPLETE.TXT"), "");
        Date validTo = jdbcTemplate.queryForObject("SELECT ValidTo FROM Organisation WHERE Navn='Rigshospitalet'", Date.class);
        Date validFrom = jdbcTemplate.queryForObject("SELECT ValidFrom FROM Organisation WHERE Navn='Rigshospitalet'", Date.class);


                                                                //2500-01-02 00:00:00.0
        Date lastValidTo = formatter.parse("2500-01-01 23:59:58");
        Date firstInvalidTo = formatter.parse("2500-01-02 00:00:01");
        assertTrue(validTo.after(lastValidTo));
        assertTrue(validTo.before(firstInvalidTo));

        Date firstValidFrom = formatter.parse("1976-04-01 00:00:01");
        Date lastInvalidBefore = formatter.parse("1976-03-31 23:59:58");
        assertTrue(validFrom.after(lastInvalidBefore));
        assertTrue(validFrom.before(firstValidFrom));
    }

    @Test
    public void updatesValidToAndModifiedDate() throws IOException, InterruptedException {

        importer.process(datasetDirWith("data/sks/SHAKCOMPLETE.TXT"), "");
        Timestamp timestamp = new Timestamp((new Date()).getTime());
        Timestamp modified1 =
                jdbcTemplate.queryForObject("SELECT ModifiedDate FROM Organisation LIMIT 1", Timestamp.class);

        // Check no records are invalidated
        long cntFirstImport = jdbcTemplate.queryForLong("SELECT count(1) FROM Organisation WHERE ValidTo<=?", timestamp);

        // Check no invalid records exist

        Thread.sleep(1000);
        importer.process(datasetDirWith("data/sks2/SHAKCOMPLETE.TXT"), "");
        timestamp = new Timestamp((new Date()).getTime());

        // Check some records have been invalidated
        long cntSecondImport = jdbcTemplate.queryForLong("SELECT count(1) FROM Organisation WHERE ValidTo<=?", timestamp);
        assertTrue(cntSecondImport > cntFirstImport);

        // Check modified date has changed
        Timestamp modified2 = jdbcTemplate.queryForObject(
                "SELECT ModifiedDate FROM Organisation ORDER BY ModifiedDate DESC LIMIT 1", Timestamp.class);
        assertFalse(modified1.equals(modified2));
    }

	private File datasetDirWith(String filename) throws IOException {
		File datasetDir = tmpDir.newFolder();
		FileUtils.copyFileToDirectory(getFile(filename), datasetDir);
		return datasetDir;
	}

	private File getFile(String filename) {
		return toFile(getClass().getClassLoader().getResource(filename));
	}
}


src/test/java/dk/nsi/sdm4/sks/integrationtest/SKSimporterIsUpIT.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dk.nsi.sdm4.sks.integrationtest;

import dk.nsi.sdm4.testutils.StatuspageChecker;
import org.junit.Test;

public class SKSimporterIsUpIT {
	@Test
	public void statusPageReturns200OK() throws Exception {
		new StatuspageChecker("sksimporter").assertThatStatuspageReturns200OK();
	}
}


src/main/java/dk/nsi/sdm4/sks/Institution.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package dk.nsi.sdm4.sks;

import dk.nsi.sdm4.core.domain.AbstractStamdataEntity;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import java.util.Date;

@Entity(name = "Organisation")
public class Institution extends AbstractStamdataEntity
{
	private Date validFrom;
	private Date validTo;
	
	private String navn;
	private String nummer;

	private final InstitutionType type;

	public enum InstitutionType
	{
		HOSPITAL_DEPARTMENT, HOSPITAL
	}

	public Institution(InstitutionType organisationstype)
	{
		this.type = organisationstype;
	}

	public Date getValidTo()
	{
		return validTo;
	}

	public void setValidTo(Date validTo)
	{
		this.validTo = validTo;
	}

	@Column
	public String getNavn()
	{
		return navn;
	}

	public void setNavn(String navn)
	{
		this.navn = navn;
	}

	@Id
	@Column
	public String getNummer()
	{
		return nummer;
	}

	public void setNummer(String nummer)
	{
		this.nummer = nummer;
	}

	@Column
	public String getOrganisationstype()
	{
		if (type == InstitutionType.HOSPITAL_DEPARTMENT)
		{
			return "Afdeling";
		}
		else if (type == InstitutionType.HOSPITAL)
		{
			return "Sygehus";
		}

		return null;
	}

	public void setValidFrom(Date validFrom)
	{
		this.validFrom = validFrom;
	}

	@Override
	public Date getValidFrom()
	{
		return validFrom;
	}
}


src/main/java/dk/nsi/sdm4/sks/config/WebConfig.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dk.nsi.sdm4.sks.config;

import dk.nsi.sdm4.core.status.StatusReporter;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.servlet.config.annotation.EnableWebMvc;

@Configuration
@EnableWebMvc
public class WebConfig {

    @Bean
    public StatusReporter statusReporter() {
        return new StatusReporter();
    }
}


src/main/java/dk/nsi/sdm4/sks/config/SksimporterApplicationConfig.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dk.nsi.sdm4.sks.config;

import dk.nsi.sdm4.core.parser.Parser;
import dk.nsi.sdm4.core.persistence.AuditingPersister;
import dk.nsi.sdm4.core.persistence.Persister;
import dk.nsi.sdm4.sks.SKSParser;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class SksimporterApplicationConfig {
    @Bean
    public Parser parser() {
		return new SKSParser();
	}

    @Bean
    public Persister persister() {
        return new AuditingPersister();
    }
}


src/main/java/dk/nsi/sdm4/sks/config/SksimporterInfrastructureConfig.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dk.nsi.sdm4.sks.config;

import dk.nsi.sdm4.core.config.StamdataConfiguration;
import dk.nsi.sdm4.core.parser.Parser;
import dk.nsi.sdm4.core.parser.ParserException;
import dk.nsi.sdm4.core.persistence.AuditingPersister;
import dk.nsi.sdm4.core.persistence.Persister;
import dk.nsi.sdm4.sks.SKSParser;
import dk.sdsd.nsp.slalog.api.SLALogConfig;
import dk.sdsd.nsp.slalog.api.SLALogger;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.scheduling.annotation.EnableScheduling;
import org.springframework.transaction.annotation.EnableTransactionManagement;

import java.io.File;
import java.sql.SQLException;

@Configuration
@EnableScheduling
@EnableTransactionManagement
//The Spring Java Configuration annotations above needs to be on this class, not on the abstract superclass to
// make Spring stop complaining about weird things
public class SksimporterInfrastructureConfig extends StamdataConfiguration {
	@Bean
	public SLALogger slaLogger() {
		return new SLALogConfig("Stamdata SKS-importer", "sksimporter").getSLALogger();
	}
}


src/main/java/dk/nsi/sdm4/sks/SKSParser.java
/**
 * The MIT License
 *
 * Original work sponsored and donated by National Board of e-Health (NSI), Denmark
 * (http://www.nsi.dk)
 *
 * Copyright (C) 2011 National Board of e-Health (NSI), Denmark (http://www.nsi.dk)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package dk.nsi.sdm4.sks;

import com.google.common.base.Preconditions;
import dk.nsi.sdm4.core.domain.Dataset;
import dk.nsi.sdm4.core.parser.Parser;
import dk.nsi.sdm4.core.parser.ParserException;
import dk.nsi.sdm4.core.persistence.Persister;
import dk.sdsd.nsp.slalog.api.SLALogItem;
import dk.sdsd.nsp.slalog.api.SLALogger;
import org.apache.commons.io.FileUtils;
import org.apache.commons.io.LineIterator;
import org.joda.time.format.DateTimeFormatter;
import org.joda.time.format.ISODateTimeFormat;
import org.springframework.beans.factory.annotation.Autowired;

import java.io.File;
import java.io.IOException;
import java.util.Calendar;
import java.util.Date;
import java.util.Iterator;

import static dk.nsi.sdm4.sks.Institution.InstitutionType.HOSPITAL;
import static dk.nsi.sdm4.sks.Institution.InstitutionType.HOSPITAL_DEPARTMENT;

/**
 * Parser for the SKS register.
 * <p/>
 * SKS is an acronym for 'Sundhedsvæsenets Klassifikationssystem'.
 * The input file is documented in the sks.pdf file in the doc directory.
 * <p/>
 * Ændringer til SKS-registeret (sgh/afd) indlæses via deltafiler.
 * Ved etablering af registeret anvendes en deltafil der indeholder
 * samtlige sgh/afd dvs. indlæsningen foretages på præcis samme måde
 * hvadenten der indlæses/opdateres et fuld register eller blot ændringer
 * siden sidst (delta)
 * <p/>
 * Eksempel på deltafil-indhold:
 * <p/>
 * afd1301011             197901011979010119821231ANÆSTHESIAFD. AN                                                                                                        084
 * afd1301011             198301011983010119941231ANÆSTESIAFD. AN,ANÆSTESIAFSNIT                                                                                          084
 * afd1301011             199501011999112419991231ANÆSTESIAFD. AN,ANÆSTESIAFSNIT                                                                                          084         SKS     3
 * afd1301011             200001012004102920031231Anæstesiologisk klinik AN, anæstesiafsnit                                                                               084         SKS     3
 * afd1301011             200401012004102925000101Anæstesi-/operationsklinik, ABD                                                                                         084         SKS     1
 * <p/>
 * Hver række angiver en sgh/afd med nummer, gyldighedperiode, navn samt
 * operationskode (3=opdatering, 1=ny eller 2=sletning)
 * Der anvendes fastpositionering dvs. værdierne er altid placeret på
 * samme position og der anvendes whitespaces til at "fylde" ud med
 * <p/>
 * Der er intet krav om at rækkefølgen for hvert nummer skal være
 * kronologisk dvs. der tages højde for at der efter at være indlæst en
 * sgh/afd med gyldighedsperiode
 * <p/>
 * 01.01.2008 - 01.01.2500
 * <p/>
 * kan optræde en anden record for samme nummer med gyldighedsperiode
 * <p/>
 * 01.01.2000 - 31.13.2007
 * <p/>
 * Det garanteres dog at der ikke optræder overlap på gyldighedsperioden
 * for samme nummer.
 * <p/>
 * Operationskoden (action) (position 187-188) angiver om recorden skal
 * betragtes som ny, opdatering eller sletning.
 * Med den måde hvorpå SKS-registeret anvendes i PEM gælder det at alle
 * entries/versioner af hvert nummer skal være placeret i
 * Organisationshistorik-tabellen (og altså ikke kun gamle versioner i
 * denne tabel) dvs. det er altid muligt heri at finde den gyldige/aktive
 * sgh/afd for en bestemt dato. I Organisations-tabellen derimod placeres
 * kun den nyeste record for en given sgh/afd dvs. recorden med nyeste
 * gyldighedsdato.
 * For at sikre denne versionering skal enhver entry (med operationskode 1
 * eller 3) altid skal indsættes/opdateres i Organisationshistorik-tabellen,
 * mens kun nyeste entry (med operationskode 1 eller 3) skal
 * indsættes/opdateres i Organisations-tabellen.
 * Det gælder at operationskode/action (1,2,3) kun er angivet for entries
 * nyere end 1995. Da vi kun ønsker at indlæse records nyere end 1995
 * ignoreres alle records hvor operationskode ikke er angivet.
 */
public class SKSParser implements Parser {
	@Autowired
	private SLALogger slaLogger;

	@Autowired
	private Persister persister;

	private static final int SKS_CODE_START_INDEX = 3;
	private static final int SKS_CODE_END_INDEX = 23;

	private static final int CODE_TEXT_START_INDEX = 47;

	/**
	 * The field is actually 120 characters long. But the
	 * specification says only to use the first 60.
	 */
	private static final int CODE_TEXT_END_INDEX = 107;

	private static final int ENTRY_TYPE_START_INDEX = 0;
	private static final int ENTRY_TYPE_END_INDEX = 3;

	private static final char OPERATION_CODE_NONE = ' ';
	private static final char OPERATION_CODE_CREATE = '1';
	private static final char OPERATION_CODE_UPDATE = '3';

	private static final String RECORD_TYPE_HOSPITAL = "sgh";
	private static final String RECORD_TYPE_DEPARTMENT = "afd";

	private static final int OPERATION_CODE_INDEX = 187;

	private static final String FILE_ENCODING = "ISO8859-15";

	private static final DateTimeFormatter dateFormat = ISODateTimeFormat.basicDate();

	protected boolean validateInputStructure(File datadir) {
		File[] input = datadir.listFiles();

		Preconditions.checkNotNull(input, "input");
		Preconditions.checkArgument(input.length > 0, "At least one file should be present at this point.");

		return input.length == 1 && isValidFilePresent(input);
	}

	private boolean isValidFilePresent(File[] input) {
		final String filename = input[0].getName();

		return filename.equalsIgnoreCase("SHAKCOMPLETE.TXT")
				|| filename.equalsIgnoreCase("SHAKDELTA.TXT");
	}

	@Override
	public void process(File datadir, String identifier) {
		Preconditions.checkState(validateInputStructure(datadir), "Input structure is invalid");

		File[] files = datadir.listFiles();

        SLALogItem slaLogItem = slaLogger.createLogItem(getHome()+".process", "SDM4."+getHome()+".process");
        slaLogItem.setMessageId(identifier);
        slaLogItem.addCallParameter(Parser.SLA_INPUT_NAME, datadir.getAbsolutePath());
		try {
			Preconditions.checkArgument(files.length == 1, "Only one file should be present at this point.");
            persister.resetTransactionTime();

            long processed = 0;
			LineIterator lines = null;
			try {
				lines = FileUtils.lineIterator(files[0], FILE_ENCODING);

				Dataset<Institution> dataset = innerParse(lines);
				persister.persistDeltaDataset(dataset);
                processed += dataset.size();
			} catch (IOException e) {
				throw new ParserException(e);
			} catch (Exception e) {
				// the persister throws these. Let's make them unchecked from here on at least
				throw new ParserException(e);
			} finally {
				LineIterator.closeQuietly(lines);
			}
            slaLogItem.addCallParameter(Parser.SLA_RECORDS_PROCESSED_MAME, ""+processed);
			slaLogItem.setCallResultOk();
			slaLogItem.store();
		} catch (RuntimeException e) {
			slaLogItem.setCallResultError("SKSParser failed - Cause: " + e.getMessage());
			slaLogItem.store();

			throw e;
		}
	}

	@Override
	public String getHome() {
		return "sksimporter";
	}

	private Dataset<Institution> innerParse(Iterator<String> lines) {
		Dataset<Institution> dataset = new Dataset<Institution>(Institution.class);

		while (lines.hasNext()) {
			Institution institution = parseLine(lines.next());

			if (institution != null) {
				dataset.add(institution);
			}
		}

		return dataset;
	}

	private Institution parseLine(String line) {
		// Determine the record type.
		//
		String recordType = line.substring(ENTRY_TYPE_START_INDEX, ENTRY_TYPE_END_INDEX);

		if (!recordType.equals(RECORD_TYPE_DEPARTMENT) && !recordType.equals(RECORD_TYPE_HOSPITAL)) {
			throw new ParserException("Unknown record type. line=" + line);
		}

		// Since the old record types do not have a operation code (and we are not
		// interested in those records) we can ignore the line.
		//
		if (line.length() < OPERATION_CODE_INDEX + 1) {
			return null;
		}

		// Determine the operation code.
		//
		char code = line.charAt(OPERATION_CODE_INDEX);

		if (code == OPERATION_CODE_CREATE || code == OPERATION_CODE_UPDATE) {
			// Create and update are handled the same way.

			Institution.InstitutionType type = recordType.equals(RECORD_TYPE_DEPARTMENT) ? HOSPITAL_DEPARTMENT : HOSPITAL;

			Institution institution = new Institution(type);

			institution.setNummer(line.substring(SKS_CODE_START_INDEX, SKS_CODE_END_INDEX).trim());

			institution.setValidFrom(dateFormat.parseDateTime(line.substring(23, 31)).toDate());
			institution.setValidTo(parseValidTo(line));

			institution.setNavn(line.substring(CODE_TEXT_START_INDEX, CODE_TEXT_END_INDEX).trim());

			return institution;
		} else if (code == OPERATION_CODE_NONE) {
			return null;
		} else {
			throw new ParserException("SKS parser encountered an unknown operation code in line " + line + ". code=" + code);
		}
	}

    private Date parseValidTo(String line) {
        // ValidTo is inclusive
        Date validToInc = dateFormat.parseDateTime(line.substring(39, 47)).toDate();
        Calendar c = Calendar.getInstance();
        c.setTime(validToInc);
        // Add a day because validTo day is inclusive
        c.add(Calendar.DATE, 1);
        return c.getTime();
    }
}
